File: src/main/scala/graveler/math/Vec3iExtensions.scala
package graveler.math

import net.minecraft.util.math.Vec3i
import scala.math._

object Vec3iExtensions {
  implicit class ExtendedVec3i(private val vec: Vec3i) {
    def +(that: Vec3i): Vec3i = {
      new Vec3i(
        vec.getX + that.getX,
        vec.getY + that.getY,
        vec.getZ + that.getZ
      )
    }

    def /(scalar: Double): Vec3i = {
      new Vec3i(vec.getX / scalar, vec.getY / scalar, vec.getZ / scalar)
    }

    def -(that: Vec3i): Vec3i = this + (-that)

    def dot(that: Vec3i): Int = {
      val x = vec.getX * that.getX
      val y = vec.getY * that.getY
      val z = vec.getZ * that.getZ

      x + y + z
    }

    def norm: Double = sqrt((vec dot vec).toDouble)

    def unary_-(): Vec3i = new Vec3i(-vec.getX, -vec.getY, -vec.getZ)
  }
}

File: src/main/kotlin/graveler/math/Vec3iExtensions.kt
package graveler.math

import kotlin.math.sqrt
import net.minecraft.util.math.Vec3i

operator fun Vec3i.unaryMinus(): Vec3i {
  return Vec3i(-x, -y, -z)
}

operator fun Vec3i.plus(b: Vec3i): Vec3i {
  return Vec3i(x + b.x, y + b.y, z + b.z)
}

operator fun Vec3i.minus(b: Vec3i): Vec3i {
  return this + (-b)
}

operator fun Vec3i.div(scalar: Double): Vec3i {
  return Vec3i(x / scalar, y / scalar, z / scalar)
}

fun Vec3i.dot(b: Vec3i): Int {
  return x * b.x + y * b.y + z * b.z
}

val Vec3i.norm: Double
  get() = sqrt(this.dot(this).toDouble())
